cirq-core/cirq/circuits/circuit_operation.py

```diff
@@ -96,6 +99,7 @@ class CircuitOperation(ops.Operation):
     param_resolver: study.ParamResolver = study.ParamResolver()
     repetition_ids: Optional[List[str]] = dataclasses.field(default=None)
     parent_path: Tuple[str, ...] = dataclasses.field(default_factory=tuple)
+    extern_keys: FrozenSet[value.MeasurementKey] = dataclasses.field(default_factory=frozenset)
```

Comment:
I see your point. I'll have a think on this one. I remember trying to do the same with parent_path: it bugged me that this was a transient thing only for decomposition, so I tried to remove the member here and push that through the protocols as needed, but it wasn't possible because of `decompose_once`. But maybe `extern_keys` is different and we can avoid storing it here. Let me ponder that a bit.

Reply:
I think that'll require explicitly passing it into mapped_circuit, decompose, and act_on. Which I think would require updating all of the implementors of that protocol with the extra parameter.

That sounds like a lot, but I'm going to see if there's some way to make that easier. The more I think about this, the more I don't like this variable being here either. It'll be empty except in partially decomposed circuits (so the concern about not being able to reuse subcircuits is probably not an issue), but it's just odd that it is there and could potentially *not match* what's in the outer circuit.

Reply:
Hmm...if it's only set on transient CircuitOperations during decomposition, I'm less opposed to it - though I think we'd still need guards in place to prevent non-transient CircuitOperations from using it. Maybe a private subclass which is identical except for this variable?

Reply:
So I don't see an easy way to get rid of this. The reason is that the decompose protocol is stateless. It just bashes things apart recursively one step at a time until some `keep` matches. It doesn't have any context that it pushes through to the things that it's bashing apart, so that transient state has to be stored somewhere. On the operation itself is the only place to store it.

Here are the options as I see them:
1. We can keep this as is (or modify it slightly as you mention above).
2. We can update decompose to thread that state through everything. (I *feel like* this would be putting too much domain-specific logic into decompose, but it may be worth trying out and seeing how the resulting logic looks).
3. We can eliminate partial decomposition of subcircuits, such that mapped_circuit is always deep. This will allow us to get rid of the `parent_path` member as well, and hopefully would simplify things in general. But if there's a requirement prohibiting this approach then we can eliminate it.

For (1) I'll admit subclassing CircuitOperation makes me feel itchy. Though maybe it works, and perhaps we could plop `parent_path` in there too. Alternately, maybe a CircuitOperationContainer that contains the circuit operation and the other things could be a viable and less itchy option.

One additional point is I think I need some more tests around classically controlled circuitoperations. Since mapped_circuit has special logic for isinstance(CircuitOperation), I'm not sure if something gets broken if the circuit operation is inside a classicallycontrolledoperation. Or even a taggedoperation for that matter. Also the existing tests should validate that decompose(op)==op.mapped_circuit(deep=True).

Anyway, is (3) an option?

Reply:
If we go for (3), we need to make sure it doesn't break this call:
https://github.com/quantumlib/Cirq/blob/e97768c6a29e1e9d2d1dcc8af86e219e82b808fd/cirq-google/cirq_google/optimizers/convert_to_sycamore_gates.py#L127-L133
Setting `preserve_structure=True` in `cirq.decompose` causes gates within a `CircuitOperation` to be decomposed without breaking the `CircuitOperation` structure. We need this in `convert_to_sycamore_gates.py` to preserve the benefits of `CircuitOperation`s in serialization; however, AFAICT there's no tests yet for using this with nested `CircuitOperation`s.

Reply:
Yeah I'm running into all kinds of bugs when nesting various permutations of circuitops with/without reps, classically controlled ops, and tagged ops. I think the ones involving layers of circuitops and taggedops were already there, though I haven't tested.

There's even kind of a requirements question, of whether `mapped_circuit(deep=True)` means to go all the way through those layers, or just sequential layers of circuitops. The current behavior of `circuitop(taggedop(circuitop)).mapped_circuit(deep=True)` is to crash, which I'm guessing is *not* the desired behavior (and I'm pretty sure this will happen on master too, though I haven't tried). Decompose seems to do the right thing though.

Anyway, it looks like this is going to take some time to sort it all out.

Reply:
*Actually* it's not all that bad. I found the crash was a simple root issue due to the `with_prefix` protocol returning NotImplemented. I will ensure the `with_scope` returns the original value if the protocol isn't implemented on it, and I think we should make a breaking change to do that on all the `with_...` protocols eventually.

So, `mapped_circuit(deep=True)` does get blocked by a TaggedOp or anything else between two circuitops, and that's the case on master branch too. So I think the best option is to make sure that behavior is unchanged (unless someone *wants* to change it, in which case we should do that in a separate PR first). Decompose, however, smashes through anything in between, which is expected.

I think (3) is my last choice, since it specifically reduces functionality that could reasonably have a use case. I'll experiment with the other two options and see where it gets us.

Another note, it appears the `preserve_structure` does not work with classical controls. It handles multiple layers of subcircuits and tags, as can be seen in `test_decompose_preserving_structure`. There's some special code in `_decompose_preserving_structure` that handles tags. So, obviously it would be nice to make that work too, preferably without special-casing everything.

Reply:
I tried (2) but it goes too deep. I'm sure it's possible but it would be a long project to do it without breaking changes.

My opinion is to keep this as-is, since that field will be empty for circuits that haven't been through a decompose_once. Otherwise I can try a subclass and see how it turns out. I could also make the field private if that helps. I also don't think my other idea of having a container would work because it breaks backwards compatibility.

Reply:
Realized this while reviewing another PR: my first comment in this chain is off the mark. It's the `FrozenCircuit` that must be invariant in order to reuse `CircuitOperation`s in multiple places.

As a result, I'm okay with keeping this as-is so long as we _thoroughly_ document its usage.

Reply:
Okay, I remember being surprised by that. I'll document further and I'll continue poking for a way to remove it.

Reply:
Done.